crates/ruff_linter/src/rules/pylint/rules/unnecessary_list_index_lookup.rs

```diff
@@ -0,0 +1,275 @@
+
+use crate::checkers::ast::Checker;
+
+/// ## What it does
+/// Checks for access of a list item at the current index when using enumeration.
+///
+/// ## Why is this bad?
+/// It is more succinct to use the variable for the value at the current index which is already in scope from the iterator.
+///
+/// ## Example
+/// ```python
+/// letters = ["a", "b", "c"]
+///
+/// for index, letter in enumerate(letters):
+///     print(letters[index])
+/// ```
+///
+/// Use instead:
+/// ```python
+/// letters = ["a", "b", "c"]
+///
+/// for index, letter in enumerate(letters):
+///     print(letter)
+/// ```
+#[violation]
+pub struct UnnecessaryListIndexLookup;
+
+impl AlwaysFixableViolation for UnnecessaryListIndexLookup {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        format!("Unnecessary lookup of list item by index")
+    }
+
+    fn fix_title(&self) -> String {
+        format!("Use existing item variable instead")
+    }
+}
+
+struct SubscriptVisitor<'a> {
+    sequence_name: &'a str,
+    index_name: &'a str,
+    diagnostic_ranges: Vec<TextRange>,
+    modified: bool,
+}
+
+impl<'a> SubscriptVisitor<'a> {
+    fn new(sequence_name: &'a str, index_name: &'a str) -> Self {
+        Self {
+            sequence_name,
+            index_name,
+            diagnostic_ranges: Vec::new(),
+            modified: false,
+        }
+    }
+}
+
+fn check_target_for_assignment(expr: &Expr, sequence_name: &str, index_name: &str) -> bool {
+    // if we see the sequence, a subscript, or the index being modified, we'll stop emitting diagnostics
+    match expr {
+        Expr::Name(ast::ExprName { id, .. }) => id == sequence_name || id == index_name,
+        Expr::Subscript(ast::ExprSubscript { value, slice, .. }) => {
+            if let Expr::Name(ast::ExprName { id, .. }) = value.as_ref() {
+                if id == sequence_name {
+                    if let Expr::Name(ast::ExprName { id, .. }) = slice.as_ref() {
+                        if id == index_name {
+                            return true;
+                        }
+                    }
+                }
+            }
+            false
+        }
+        _ => false,
+    }
+}
+
+impl<'a> Visitor<'_> for SubscriptVisitor<'a> {
+    fn visit_expr(&mut self, expr: &Expr) {
+        if self.modified {
+            return;
+        }
+        match expr {
+            Expr::Subscript(ast::ExprSubscript {
+                value,
+                slice,
+                range,
+                ..
+            }) => {
+                if let Expr::Name(ast::ExprName { id, .. }) = value.as_ref() {
+                    if id == self.sequence_name {
+                        if let Expr::Name(ast::ExprName { id, .. }) = slice.as_ref() {
+                            if id == self.index_name {
+                                self.diagnostic_ranges.push(*range);
+                            }
+                        }
+                    }
+                }
```

Comment:
nit: can we avoid having nested `if` conditions by using `let Some(ast::ExprName { id, .. }) = value.as_name_expr() else { return; }`